plugin/access.lua

```diff
@@ -14,170 +37,223 @@ local function array_has_value(arr, val)
--- We use the access_token value as the cache key
--- Note also that we pass two arguments to the introspect_access_token callback
+-- Get the token from the cache or introspect it
 --
 local function verify_access_token(access_token, config)
-    
-    local res, err = kong.cache:get(
-        access_token,
-        { ttl = config.token_cache_seconds },
-        introspect_access_token,
-        access_token,
-        config)
-
-    if err then
-        local cacheMessage = "A technical problem occurred during cache access"
-        ngx.log(ngx.WARN, cacheMessage .. err)
-        error_response(ngx.HTTP_INTERNAL_SERVER_ERROR, "cache_error", cacheMessage, config)
-    end 
 
-    if res.status ~= 200 then
-        kong.cache:invalidate(access_token)
+    -- Return previous introspeciton results for the same token if available
+    local dict = ngx.shared['phantom-token']
+    local existing_jwt = dict:get(access_token)
+    if existing_jwt then
+        return { status = 200, jwt = existing_jwt }
     end
 
-    if res.status == 204 then
-        ngx.log(ngx.WARN, "Received a " .. res.status .. " introspection response due to the access token being invalid or expired")
-        invalid_token_error_response(config)
-    end
+    -- Otherwise introspect the opaque access token
+    local result = introspect_access_token(access_token, config)
+    if result.status == 200 then
 
-    if res.status ~= 200 then
+        if not verify_scope(result.jwt, config.scope) then
+            return { status = 403 }
+        end
 
-        local introspectionMessage = "A technical problem occurred during access token introspection"
-        local logMessage = introspectionMessage .. ", Status: " .. res.status
-        if res.body then
-            logMessage = logMessage .. ", Body: " .. res.body
+        local time_to_live = config.token_cache_seconds
+        if result.expiry > 0 and result.expiry < config.token_cache_seconds then
+            time_to_live = result.expiry
         end
-        ngx.log(ngx.WARN, logMessage)
 
-        error_response(ngx.HTTP_INTERNAL_SERVER_ERROR, "introspection_error", introspectionMessage, config)
+        -- Cache the result so that introspection is efficient under load
+        -- The opaque access token is already a unique string similar to a GUID so use it as a cache key
+        -- The cache is atomic and thread safe so is safe to use across concurrent requests
+        -- The expiry value is a number of seconds from the current time
+        -- https://github.com/openresty/lua-nginx-module#ngxshareddictset
+        dict:set(access_token, result.jwt, time_to_live)
     end
 
-    return res
+    return result
 end
 
 --
 -- The public entry point to introspect the token then forward the JWT to the API
 --
 function _M.run(config)
 
-    if ngx.req.get_method() == "OPTIONS" then
+    -- Start by validating configuration
+    if initialize_configuration(config) == false then 
```

Comment:
Is there any way of making those checks part of the Nginx startup? It seems a bit excessive to be doing that on every request. Also not sure if it's good practice to have side effects on the `config` variable in this method call.

Reply:
I saw in the tests that we deliberately don't want Nginx to crash if it's misconfigured. I wonder which is better from a performance perspective.

Reply:
C modules can crash at startup if the config is wrong, but I don't think there is a way to do this in LUA, which is why I used this approach. I have added a future issue to think more about this, but for now this repo is in line with the OAuth Proxy plugin, which has the same behavior.